priv/posts/08-diagnostics-channel-is-landed.md

```diff
@@ -0,0 +1,105 @@
+With `diagnostics_channel`, Node.js core and module authors can publish contextual data about what they are doing at a given time. This could be the hostname and query string of a mysql query, for example. Just create a named channel with dc.channel(name) and call channel.publish(data) to send the data to any listeners to that channel._
+
+This feature is similar to [`EventEmmiter`](https://nodejs.org/api/events.html#events_class_eventemitter), however is optimized enough to propagate data synchronous.
+
+APM Vendors usually does `monkey-patch` of every key modules for publish information. In `diagnostics_channel` world, we can avoid it in favor of events.
+
+## Usage
+
+That is a simple api, let's see some examples:
+
+This one is a subscriber to `root.caughtError` channel.
+
+```js
+const dc = require('diagnostics_channel')
+const channel = dc.channel('root.caughtError')
+
+channel.subscribe({ error }) => {
+  console.error('One error was propagated through dc', error)
+})
+```
+
+> But, when it will be called?
+
+Well, in some part of your application (even libraries used) this event should be called:
+
+```js
+const dc = require('diagnostics_channel')
+const channel = dc.channel('root.caughtError')
+
+async function exampleFunction() {
+  try {
+    await throwableFunction()
+  } catch (e) {
+    channel.publish({ error: e })
+  }
+}
+```
+
+Or, let's do a more reasonable example, we want to measure time elapsed of each query performed:
+
+```js
+// module mysql.js
+const dc = require('diagnostics_channel')
+const channel = dc.channel('mysql.query')
+
+MySQL.prototype.query = async function query(queryString, values, callback) {
+  const start = Date.now() // You can do it with perf_hooks as well
+  await this.doQuery(queryString, values, callback);
+  const end = Date.now()
+
+  // Broadcast query information whenever a query is done
+  channel.publish({
+    query: queryString,
+    host: this.hostname,
+    timeElapsed: end - start,
+  })
+}
+```
+
+It able us to measure bottlenecks in our code:
+
+```js
+const dc = require('diagnostics_channel')
+const channel = dc.channel('mysql.query')
+
+channel.subscribe({ timeElapsed, query } => {
+  if (timeElapsed > process.env.QUERY_THRESHOLD) {
+    console.warn('Query slow: ', query, timeElapsed)
+  }
+})
+```
+
+Actually, most of the database modules already emit the warning on query slow if you set on the settings. This example is just to show you the main usage of this module.
+
+After key modules support `diagnostics_channel` we are able to have a better observability/tracing of our application without add a lot of complexity in our nodejs code.
+
+For instance, we can have:
+
+- Tracking entire http lifecycle
+- Measure core metrisc of our application easier (event-loop, gc, memory utilization, cpu utilization)
+
+Of course, this feature sounds better for APM Vendors.
+
+## Why use this module instead of EventEmmiter?
+
+// TODO
```

Comment:
@Qard, can you help here?

Reply:
`EventEmitter` has an extra cost on every single publish to look up the handler set by the string event name. That's not a huge deal for just a single run, but in a high-frequency scenario where the logic might be repeated thousands, or even _millions_, of times per second, it adds up _fast_. Additionally, the lookup cost _always_ happens with `EventEmitter` while with `diagnostics_channel` it _only_ happens when something is actually _listening_ to that specific channel. The intent is for there to be _hundreds_ or even _thousands_ of these channels being reported to at any given time while there might be only a _few_ of those channels being actively _observed_ at any given time. The majority of the time there would be nothing to publish to so it's been intentionally designed to do _nothing at all_ in that case. This design makes it much more suitable as a _data firehose_ whereas a typical `EventEmitter` is really only suited to a more _limited_ set of events.